hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/upgrade/TestDatanodeUpgradeToSchemaV4.java

```diff
@@ -0,0 +1,383 @@
+    File yamlFile = container.getContainerFile();
+    String content =
+        FileUtils.readFileToString(yamlFile, Charset.defaultCharset());
+    System.out.println(content);
+    // yaml file contains chunkPath and metadataPath
+    assertThat(content).contains(METADATA_PATH);
+    assertThat(content).contains(CHUNKS_PATH);
+    assertEquals(yamlFile.getParentFile().getParentFile().toPath().resolve(STORAGE_DIR_CHUNKS).toString(),
+        container.getContainerData().getChunksPath());
+    assertEquals(yamlFile.getParentFile().getAbsolutePath(), container.getContainerData().getMetadataPath());
+    File containerDir = new File(container.getContainerData().getContainerPath());
+    assertTrue(containerDir.exists() && containerDir.isDirectory());
+    FileTime creationTime1 = (FileTime) Files.getAttribute(containerDir.toPath(), "creationTime");
+
+    File folderToExport = Files.createFile(
+        tempFolder.toPath().resolve("export-testContainerBeforeFinalization.tar")).toFile();
+    TarContainerPacker packer = new TarContainerPacker(NO_COMPRESSION);
+    if (exportBeforeFinalization) {
+      //export the container
+      try (FileOutputStream fos = new FileOutputStream(folderToExport)) {
+        container.exportContainerData(fos, packer);
+      }
+    }
+
+    dsm.finalizeUpgrade();
+
+    if (!exportBeforeFinalization) {
+      //export the container
+      try (FileOutputStream fos = new FileOutputStream(folderToExport)) {
+        container.exportContainerData(fos, packer);
+      }
+    }
+
+    //delete the original one
+    KeyValueContainerUtil.removeContainer(container.getContainerData(), conf);
+    container.delete();
+    assertFalse(new File(container.getContainerData().getContainerPath()).exists());
+    if (schemaV3Enabled) {
+      assertTrue(container.getContainerData().getDbFile().exists());
+    }
+
+    //create a new one
+    KeyValueContainerData oldContainerData = container.getContainerData();
+    KeyValueContainerData newContainerData =
+        new KeyValueContainerData(containerID1,
+            oldContainerData.getLayoutVersion(),
+            oldContainerData.getMaxSize(), pipeline.getId().getId().toString(),
+            dsm.getDatanodeDetails().getUuidString());
+    newContainerData.setSchemaVersion(oldContainerData.getSchemaVersion());
+    KeyValueContainer newContainer = new KeyValueContainer(newContainerData, conf);
+    newContainer.populatePathFields(scmTestMock.getClusterId(), oldContainerData.getVolume());
+
+    // verify yaml file checksum
+    try (FileInputStream fis = new FileInputStream(folderToExport)) {
+      byte[] containerDescriptorYaml = packer.unpackContainerDescriptor(fis);
+      KeyValueContainerData data = (KeyValueContainerData) ContainerDataYaml
+          .readContainer(containerDescriptorYaml);
+      ContainerUtils.verifyChecksum(data, conf);
+    }
+
+    // sleep 1s to make sure creationTime will change
+    Thread.sleep(1000);
```

Comment:
Can we avoid this?  Not only does it make tests slower, assertions about timing are flaky when relying on sleep.

If prod code already uses `Clock`, we can use `TestClock` in the tests.  Otherwise, I'd propose removing assertions about `creationTime`.

Reply:
The sleep here is for detecting the container directory is recreated during container import.  Without this sleep, the creationTime of old directory and new directory most time will be same. I prefer to keep the creationTime here.